pkg/actors/actors.go

```diff
@@ -299,32 +300,97 @@ func constructCompositeKey(keys ...string) string {
 	return strings.Join(keys, daprSeparator)
 }
 
-func (a *actorsRuntime) deactivateActor(actorType, actorID string) error {
-	req := invokev1.NewInvokeMethodRequest("actors/"+actorType+"/"+actorID).
-		WithActor(actorType, actorID).
+// Halts an actor, removing it from the actors table and then deactivating it
+func (a *actorsRuntime) haltActor(actorType, actorID string) error {
+	key := constructCompositeKey(actorType, actorID)
+	log.Debugf("Halting actor '%s'", key)
+
+	// Remove the actor from the table
+	// This will forbit more state changes
+	actAny, ok := a.actorsTable.LoadAndDelete(key)
+
+	// If nothing was loaded, the actor was probably already deactivated
+	if !ok || actAny == nil {
+		return nil
+	}
+
+	act := actAny.(*actor)
+	for {
+		// wait until actor is not busy, then deactivate
+		if !act.isBusy() {
+			break
+		}
+
+		a.clock.Sleep(time.Millisecond * 100)
+	}
+
+	return a.deactivateActor(act)
+}
+
+// Halts all actors
+func (a *actorsRuntime) haltAllActors() error {
+	// Visit all currently active actors and deactivate them
+	errCh := make(chan error)
+	count := atomic.Int32{}
+	a.actorsTable.Range(func(key any, value any) bool {
+		count.Add(1)
+		go func(key any) {
+			actorKey := key.(string)
+			err := a.haltActor(a.getActorTypeAndIDFromKey(actorKey))
+			if err != nil {
+				errCh <- fmt.Errorf("failed to deactivate actor '%s': %v", actorKey, err)
+			}
+			errCh <- nil
+		}(key)
+		return true
+	})
+
+	// Collect all errors, which also waits for all goroutines to return
+	errs := []error{}
+	for i := int32(0); i < count.Load(); i++ {
+		err := <-errCh
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	return errors.Join(errs...)
+}
+
+func (a *actorsRuntime) deactivateActor(act *actor) error {
+	ctx := context.Background()
```

Comment:
This method always needs to use a background context. We can cascade it from a caller, but eventually it will need to be context.Background regardless

Reply:
Why will it need to be context.Background regardless?

Reply:
`deactivateActor` is called only by `haltActor`.

`haltActor` is called by:

- `drainRebalancedActors` -> this is invoked when there's a rebalancing, so it's a background operation with a bg context
- `haltAllActors` -> this is invoked by placement when all actors need to be halted (for example when app is unhealthy), so another background operation with no context to cascade
- The ticker in `deactivationTicker` -> runs in background

There's no API that allows users to deactivate actors themselves.

Reply:
Thanks :+1: